### PR TITLE
Add support for Rails 5.1 and 5.2

### PR DESCRIPTION
**Pull request by uasi**

### CODE REVIEW

**Review by s4ichi — APPROVED** [review]

File: lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb [hunk 1]
```diff
@@ -27,29 +24,56 @@ def belongs_to_counter_cache_after_logical_delete(reflection)
...
+                    def saved_change_to_logical_delete_column?(delete_column)
+                      __send__("#{delete_column}_changed?")
+                    end
+
+                    def logical_delete_column_before_last_save(delete_column)
+                      __send__("#{delete_column}_was")
+                    end
+                  else
+                    def saved_change_to_logical_delete_column?(delete_column)
+                      __send__("saved_change_to_#{delete_column}")
```
Comment (riseshia) [inline lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb:44]: `attribute_changed?` seems to be renamed to `will_save_change_to_attribute?`, is this intended change?
https://github.com/rails/rails/pull/25337/files#diff-60c25306d36b9a651f87d1c4bb381b33be05620f6fc17ec2ddb77b2ba0011681R224
I want some explanation about this 🤔

File: lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb [hunk 2]
```diff
@@ -27,29 +24,56 @@ def belongs_to_counter_cache_after_logical_delete(reflection)
...
+                    end
+
+                    def logical_delete_column_before_last_save(delete_column)
```
Comment (riseshia) [inline lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb:47]: `attribute_was` seems to be renamed to `attribute_in_database`, is this intended change?
https://github.com/rails/rails/pull/25337/files#diff-60c25306d36b9a651f87d1c4bb381b33be05620f6fc17ec2ddb77b2ba0011681R214
I want some explanation about this 🤔

File: lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb (hunk 1 above)
Comment (uasi) [reply lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb:44]: `attribute_changed?` has been split into two methods, `will_save_change_to_attribute?` and `saved_change_to_attribute?`. See how the behavior of these methods in before/after callbacks changes on different ActiveRecord versions.

```ruby
# attr.rb

require 'rspec'
require 'spec_helper'

puts "# activerecord #{ActiveRecord.version}"
puts

def print_attr_states(at, r)
  ar_5_0 = ActiveRecord.version.to_s < '5.1.0'
  puts <<~END
    # #{at}
    attribute_changed?             : #{r.attribute_changed?(:deleted_at).inspect}
    will_save_change_to_attribute? : #{ar_5_0 ? '(undef)' : r.will_save_change_to_attribute?(:deleted_at).inspect}
    saved_change_to_attribute?     : #{ar_5_0 ? '(undef)' : r.saved_change_to_attribute?(:deleted_at).inspect}

  END
end

Logical.before_update { |r| print_attr_states('before_update', r) }
Logical.after_update { |r| print_attr_states('after_update', r) }

r = Logical.create
r.soft_delete
```

```
% bundle exec ruby -I spec attr.rb
# activerecord 5.0.7.2

# before_update
attribute_changed?             : true
will_save_change_to_attribute? : (undef)
saved_change_to_attribute?     : (undef)

# after_update
attribute_changed?             : true
will_save_change_to_attribute? : (undef)
saved_change_to_attribute?     : (undef)
```

```
% bundle exec ruby -I spec attr.rb
# activerecord 5.1.7

# before_update
attribute_changed?             : true
will_save_change_to_attribute? : true
saved_change_to_attribute?     : false

DEPRECATION WARNING: The behavior of `attribute_changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute?` instead. (called from print_attr_states at attr.rb:11)
# after_update
attribute_changed?             : true
will_save_change_to_attribute? : nil
saved_change_to_attribute?     : true
```

```
% bundle exec ruby -I spec attr.rb
# activerecord 5.2.6

# before_update
attribute_changed?             : true
will_save_change_to_attribute? : true
saved_change_to_attribute?     : false

# after_update
attribute_changed?             : false
will_save_change_to_attribute? : nil
saved_change_to_attribute?     : true
```

Regarding guard_against_physical_delete, we need to know if the logical delete attribute have been changed in the after_update callback. There, `will_save_change_to_attribute?` is designed to return nil from the beginning, and `attribute_changed?` always returns false regardless of changes starting from AR 5.2. So `saved_change_to_attribute?` is the only option.

File: lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb (hunk 2 above)
Comment (uasi) [reply lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb:47]: Like I said above, we need to know what was the persisted value before the last save in the after_update callback. This is only possible with `attribute_before_last_save`.

```ruby
require 'rspec'
require 'spec_helper'

puts "# activerecord #{ActiveRecord.version}"
puts

def print_attr_states(at, r)
  ar_5_0 = ActiveRecord.version.to_s < '5.1.0'
  puts <<~END
    # #{at}
    attribute_was              : #{r.attribute_was(:deleted_at).inspect}
    attribute_in_database      : #{ar_5_0 ? '(undef)' : r.attribute_in_database(:deleted_at).inspect}
    attribute_before_last_save : #{ar_5_0 ? '(undef)' : r.attribute_before_last_save(:deleted_at).inspect}

  END
end

# Set non-nil initial value to make it clear that `attribute_before_last_save` always return nil in the before_update callback.
r = Logical.create
r.deleted_at = Time.at(0)
r.save

Logical.before_update { |r| print_attr_states('before_update', r) }
Logical.after_update { |r| print_attr_states('after_update', r) }

r.deleted_at = Time.now
r.save
```

```
% bundle exec ruby -I spec attr_was.rb
# activerecord 5.0.7.2

# before_update
attribute_was              : 1970-01-01 00:00:00 UTC
attribute_in_database      : (undef)
attribute_before_last_save : (undef)

# after_update
attribute_was              : 1970-01-01 00:00:00 UTC
attribute_in_database      : (undef)
attribute_before_last_save : (undef)
```

```
% bundle exec ruby -I spec attr_was.rb
# activerecord 5.1.7

# before_update
attribute_was              : 1970-01-01 00:00:00 UTC
attribute_in_database      : 1970-01-01 00:00:00 UTC
attribute_before_last_save : nil

DEPRECATION WARNING: The behavior of `attribute_was` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `attribute_before_last_save` instead. (called from print_attr_states at attr_was.rb:11)
DEPRECATION WARNING: The behavior of `attribute_changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute?` instead. (called from print_attr_states at attr_was.rb:11)
DEPRECATION WARNING: The behavior of `changed_attributes` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_changes.transform_values(&:first)` instead. (called from print_attr_states at attr_was.rb:11)
# after_update
attribute_was              : 1970-01-01 00:00:00 UTC
attribute_in_database      : 2021-06-15 10:41:01.792418 UTC
attribute_before_last_save : 1970-01-01 00:00:00 UTC
```

```
% bundle exec ruby -I spec attr_was.rb
# activerecord 5.2.6

# before_update
attribute_was              : 1970-01-01 00:00:00 UTC
attribute_in_database      : 1970-01-01 00:00:00 UTC
attribute_before_last_save : nil

# after_update
attribute_was              : 2021-06-15 10:41:04.842825 UTC
attribute_in_database      : 2021-06-15 10:41:04.842825 UTC
attribute_before_last_save : 1970-01-01 00:00:00 UTC
```

File: lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb (hunk 1 above)
Comment (uasi) [inline lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb:44]: 🤭 

```suggestion
                      __send__("saved_change_to_#{delete_column}?")
```

**Review by riseshia — APPROVED** [review]
:100: :bow:

**Comment by uasi** [comment]
Thank you!

**Pull request merged**